tap_zendesk/__init__.py

```diff
@@ -117,16 +126,43 @@ def do_sync(client, catalog, state, start_date):
     singer.write_state(state)
     LOGGER.info("Finished sync")
 
+def oauth_auth():
+    try:
+        oauth_args = singer.utils.parse_args(OAUTH_CONFIG_KEYS)
+        creds = {
+            "subdomain": args['subdomain'],
+            "oauth_token": oauth_args.config['access_token'],
+        }
+
+        client = Zenpy(**creds)
+        LOGGER.info("OAuth authentication successful.")
+        return client
+    except ZenpyException:
+        LOGGER.error("OAuth authentication failed.")
+    except Exception:
+        LOGGER.error("Cannot find OAuth configuration.")
+
+def api_token_auth():
+    try:
+        api_args = singer.utils.parse_args(API_CONFIG_KEYS)
+        creds = {
+            "subdomain": api_args.config['subdomain'],
+            "email": api_args.config['email'],
+            "token": api_args.config['api_token']
+        }
+
+        client = Zenpy(**creds)
+        LOGGER.info("API Token authentication successful.")
+        return client
+    except ZenpyException:
+        LOGGER.error("API Token authentication failed.")
+    except Exception:
+        LOGGER.error("Cannot find API Token configuration.")
 
 @singer.utils.handle_top_exception(LOGGER)
 def main():
     parsed_args = singer.utils.parse_args(REQUIRED_CONFIG_KEYS)
-
-    creds = {
-        "subdomain": parsed_args.config['subdomain'],
-        "oauth_token": parsed_args.config['access_token'],
-    }
-    client = Zenpy(**creds)
+    client = oauth_auth() or api_token_auth()
```

Comment:
Nice. :)

I like the idea of dispatching on the config here rather than relying on Exceptions. Can't we verify A) Only one or the other config keys are present and B) That the one presented succeeded?

I'm mainly trying not to have that `except Exception` above and to not attempt `oauth` if `api_token` is present or `api_token` if `oauth` is present. The logs would be a bit cleaner.

Reply:
Right I can do that.